src/Microsoft.AspNetCore.Razor.Tools/GenerateCommand.cs

```diff
@@ -0,0 +1,204 @@
+
+        protected override Task<int> ExecuteCoreAsync()
+        {
+            var result = ExecuteCore(
+                projectDirectory: ProjectDirectory.Value() ?? Environment.CurrentDirectory,
+                outputDirectory: OutputDirectory.Value(),
+                tagHelperManifest: TagHelperManifest.Value(),
+                sources: Sources.Values.ToArray());
+
+            return Task.FromResult(result);
+        }
+
+        protected override bool ValidateArguments()
+        {
+            if (string.IsNullOrEmpty(OutputDirectory.Value()))
+            {
+                Error.WriteLine($"{OutputDirectory.ValueName} not specified.");
+                return false;
+            }
+
+            if (Sources.Values.Count == 0)
+            {
+                Error.WriteLine($"{Sources.Name} should have at least one value.");
+                return false;
+            }
+
+            return true;
+        }
+
+        private int ExecuteCore(string projectDirectory, string outputDirectory, string tagHelperManifest, string[] sources)
+        {
+            tagHelperManifest = Path.Combine(projectDirectory, tagHelperManifest);
+            outputDirectory = Path.Combine(projectDirectory, outputDirectory);
+
+            var tagHelpers = GetTagHelpers(tagHelperManifest);
+
+            var engine = RazorEngine.Create(b =>
+            {
+                RazorExtensions.Register(b);
+
+                b.Features.Add(new StaticTagHelperFeature() { TagHelpers = tagHelpers, });
+            });
+
+            var templateEngine = new MvcRazorTemplateEngine(engine, RazorProject.Create(projectDirectory));
+
+            var sourceItems = GetRazorFiles(projectDirectory, sources);
+            var results = GenerateCode(templateEngine, sourceItems);
+
+            var success = true;
+
+            foreach (var result in results)
+            {
+                if (result.CSharpDocument.Diagnostics.Count > 0)
+                {
+                    success = false;
+                    foreach (var error in result.CSharpDocument.Diagnostics)
+                    {
+                        Console.Error.WriteLine(error.ToString());
+                    }
+                }
+
+                var outputFilePath = Path.Combine(outputDirectory, Path.ChangeExtension(result.ViewFileInfo.ViewEnginePath.Substring(1), ".cs"));
+                File.WriteAllText(outputFilePath, result.CSharpDocument.GeneratedCode);
+            }
+
+            return success ? 0 : -1;
+        }
+
+        private IReadOnlyList<TagHelperDescriptor> GetTagHelpers(string tagHelperManifest)
+        {
+            if (!File.Exists(tagHelperManifest))
+            {
+                return Array.Empty<TagHelperDescriptor>();
+            }
+
+            using (var stream = File.OpenRead(tagHelperManifest))
+            {
+                var reader = new JsonTextReader(new StreamReader(stream));
+
+                var serializer = new JsonSerializer();
```

Comment:
We've done this so many times we might want to consider creating a `RazorJsonTextSerializer` that does all this.

Reply:
Or we could make them `ISerializable` and use `BinarySerializer`. I don't super want to use JSON.NET here

Reply:
We could also consider making these JSON serializers into a shared-source project so that it's easier to maintain

Reply:
Will address as part of https://github.com/aspnet/Razor/issues/1953